Hari's changes attempt 2

adampbeardsley opened the pull request:
Trying again with consolidated commits

HaraKrish committed:
New blocks in py3 bifrost on commensal machine

HaraKrish committed:
bug fixes to optimized blocks of EPIC

HaraKrish committed:
Bug fix to auto-correlation estimation block in EPIC

HaraKrish committed:
Bug fixes

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,558 @@
+
+          if (!(myGridU == grid_point_u && myGridV == grid_point_v)) 
+	  //if(!((myGridU<grid_point_u+0.5)&&(myGridU>=grid_point_u-0.5)&&(myGridV<grid_point_v+0.5)&&(myGridU>=grid_point_u-0.5)))
+	   { // Atomic add to grid is now removed from this kernel after careful examining and consideration for race-conditions
```
TODO: confirm this

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,558 @@
+
+/* 
+
+Implements the Romein convolutional algorithm onto a GPU using CUDA. 
```
Does this actually implement romein? I thought we diverged from that.

adampbeardsley commented on src/XGrid.cu:
```diff
@@ -0,0 +1,298 @@
+/* 
+
+Implements the grid-Correlation onto a GPU using CUDA. 
```
how is this different from GMul?

HaraKrish replied:
We do not need GMul, it is no longer needed.

adampbeardsley commented on src/aCorr.cu:
```diff
@@ -0,0 +1,372 @@
+
+ 
+
+Implements the grid-Correlation onto a GPU using CUDA. 
```
I'm assuming this is incorrect?

adampbeardsley replied:
Please add a meaningful description

adampbeardsley commented on src/aCorr.cu:
```diff
@@ -0,0 +1,372 @@
+};
+template<typename RealType>
+__host__ __device__
+inline Complex<RealType> ComplexMul(Complex<RealType> x, Complex<RealType> y, Complex<RealType> d) {
```
This function appears in many blocks. Can it go in a utility file?

adampbeardsley commented on src/fft_shift.cu:
```diff
@@ -0,0 +1,80 @@
+    int ix0 = y * size + x;
+    int ix1 = (ix0 + (size + 1) * (size/2)) % (size*size);
+
+    T temp = grid[ix0];
+    grid[ix0] = grid[ix1];
+    grid[ix1] = temp;
```
Does this only shift in one dimension? It looks like it only shifts x, and I don't see where y gets shifted.

HaraKrish replied:
No idea, I did not write this. May be James has an answer. We had to manually copy stuff from Jame's repo as per the earlier install instructions. I guess it just stayed

jaycedowell replied:
`fft_shift.cu` doesn't seem to be in ledatelescope bifrost.  Maybe this isn't needed and should be removed?

adampbeardsley commented on src/formats/chips.hpp:
```diff
@@ -142,7 +142,7 @@ class CHIPSProcessor : virtual public PacketProcessor {
 	    */
 	    //if( pkt->src < 8 ) { // HACK TESTING
 	    //for( ; chan<32; ++chan ) { // HACK TESTING
-	    for( ; chan<pkt->nchan; ++chan ) { // HACK TESTING
+	    for( ; chan<32; ++chan ) { // HACK TESTING pkt->nchan
```
Clean up. Why are we only unpacking 32 channels? Should this be a parameter passed in, instead of hard coded? On slack @HaraKrish said he was unpacking 132 channels, but this looks like it's only 32.

HaraKrish replied:
I'm unpacking 90 channels, this one is out-of-date.

jaycedowell replied:
I still find it strange that we have to limit how many channels are unpacked to get things running.

adampbeardsley commented on src/romein.cu:
```diff
@@ -1,5 +1,5 @@
 /*
```
Are the changes in this file still relevant, or are these the changes that eventually became the other blocks? If the latter, we should revert this to the version that really is Romein.

adampbeardsley commented on src/romein_origin.cu:
```diff
@@ -0,0 +1,641 @@
+/*
```
What is this file?

HaraKrish replied:
I just realize that this repo is not updated. Let me update a couple of files and remove them before we go ahead with this pull request. This is the original romein kernel, but we could remove this and keep romein.cu which is just a clean up version of the original romein

adampbeardsley reviewed (commented):
I took a quick pass at high-level stuff here. @HaraKrish could you take a look at my comments and see what should be moved toward the main branch? We merged ibverbs support in, so we should be able to get your stuff in next.

HaraKrish commented on src/GMul.cu:
```diff
@@ -0,0 +1,307 @@
+        extern  __shared__ Complex<float> shared[] ;
+        In* xx = reinterpret_cast<In *>(shared);
+        In* yy = xx + blk_x;
+	int tt = 1;
+        if(npol>1) tt=(int)npol/2;
+        int bid  = ((bid_x * grid_y + bid_y) * tt  * grid_z  + bid_z) * blk_x ;
+	int bid2 = ((bid_x * grid_y + bid_y) * npol * grid_z  + bid_z) * blk_x ;
+        //int bid  = (bid_x * tt  * grid_z  + bid_z) * blk_x ;
+        //int bid2 = (bid_x * grid_y * grid_z  + bid_z) * blk_x ;
```
This file isn't relevant anymore, I need to remove this

HaraKrish committed:
Remove irrelevant files

HaraKrish committed:
Updated kernels

HaraKrish committed:
Remove unused file

HaraKrish committed:
Remove XCorr.py

jaycedowell commented on src/formats/chips.hpp:
```diff
@@ -159,8 +159,8 @@ class CHIPSProcessor : virtual public PacketProcessor {
-		    for( int c=0; c<nchan; ++c ) {
-			    ::memset(&aligned_data[src + nsrc*(c + nchan*t)],
+		    for( int c=0; c<128; ++c ) {  
+			    ::memset(&aligned_data[src + nsrc*(c + 128*t)],
```
I think these should still be `nchan` in the missing source blanker.

adampbeardsley replied:
@HaraKrish will test leaving it at `nchan` (and above on line 145)

HaraKrish replied:
> @HaraKrish will test leaving it at `nchan` (and above on line 145)

This is done

jaycedowell commented on config.mk:
```diff
@@ -18,11 +18,11 @@ else
 ifndef INSTALL_LIB_DIR
-	INSTALL_LIB_DIR = /usr/local/lib
+	INSTALL_LIB_DIR = /home/epic/anaconda/lib/
 endif
 
 ifndef INSTALL_INC_DIR
-	INSTALL_INC_DIR = /usr/local/include
+	INSTALL_INC_DIR = /home/epic/anaconda/include/
```
What about those in `user.mk`?

adampbeardsley replied:
We've resolved to just accept this.

adampbeardsley committed:
Delete fft_shift.cu

adampbeardsley committed:
Delete fft_shift.h

adampbeardsley committed:
Delete fft_shift.py

adampbeardsley committed:
remove fft_shift from makefile

jaycedowell commented:
@HaraKrish are there unit tests for the new modules?

HaraKrish commented:
> @HaraKrish are there unit tests for the new modules?

No, there are no unit tests for these modules.

jaycedowell commented on src/aCorr.cu:
```diff
@@ -0,0 +1,358 @@
+   void execute(BFarray const* in, BFarray const* out, int nbatch, int nchan) {
+        BF_TRACE();
+        BF_TRACE_STREAM(_stream);
+        BF_ASSERT_EXCEPTION(out->dtype == BF_DTYPE_CF32 \
+                                          || BF_DTYPE_CF64, BF_STATUS_UNSUPPORTED_DTYPE);
```
This looks like a bug.

jaycedowell replied:
```suggestion
        BF_ASSERT_EXCEPTION((out->dtype == BF_DTYPE_CF32) \
                                          || (out->dtype == BF_DTYPE_CF64), BF_STATUS_UNSUPPORTED_DTYPE);
```

jaycedowell commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+    void set_kernels(BFarray const* kernels) {
+        BF_TRACE();
+        BF_TRACE_STREAM(_stream);
+        BF_ASSERT_EXCEPTION(kernels->dtype == BF_DTYPE_CF32 \
+                                              || BF_DTYPE_CF64, BF_STATUS_UNSUPPORTED_DTYPE);
```
This looks like a bug.

jaycedowell replied:
```suggestion
        BF_ASSERT_EXCEPTION((kernels->dtype == BF_DTYPE_CF32) \
                                              || (kernels->dtype == BF_DTYPE_CF64), BF_STATUS_UNSUPPORTED_DTYPE);
```

jaycedowell commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+        BF_ASSERT_EXCEPTION(_kernels != NULL, BF_STATUS_INVALID_STATE);
+        BF_ASSERT_EXCEPTION(out->dtype == BF_DTYPE_CF32 \
+                                          || BF_DTYPE_CF64, BF_STATUS_UNSUPPORTED_DTYPE);
+        
```
This looks like a bug.

jaycedowell commented on src/XGrid.cu:
```diff
@@ -0,0 +1,260 @@
+   void execute(BFarray const* in, BFarray const* out, int nbatch, int nchan, int npol) {
+        BF_TRACE();
+        BF_TRACE_STREAM(_stream);
+        BF_ASSERT_EXCEPTION(out->dtype == BF_DTYPE_CF32 \
+                                          || BF_DTYPE_CF64, BF_STATUS_UNSUPPORTED_DTYPE);
```
This looks like a bug.

jaycedowell replied:
```suggestion
        BF_ASSERT_EXCEPTION((out->dtype == BF_DTYPE_CF32) \
                                          || (out->dtype == BF_DTYPE_CF64), BF_STATUS_UNSUPPORTED_DTYPE);
```

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+                d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].x+= sum.x;
+                d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].y+= sum.y;
+      		    
+//              atomicAdd(&d_out[grid_s + pol*gridsize*gridsize + gridsize*grid_point_v + grid_point_u].x, sum.x);
+//              atomicAdd(&d_out[grid_s + pol*gridsize*gridsize + gridsize*grid_point_v + grid_point_u].y, sum.y);
```
remove unused code

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+                       d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].x+= sum.x;
+                       d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].y+= sum.y;   
+                   
+		   // atomicAdd(&d_out[grid_s + pol*gridsize*gridsize + gridsize*grid_point_v + grid_point_u].x, sum.x);
+                   // atomicAdd(&d_out[grid_s + pol*gridsize*gridsize + gridsize*grid_point_v + grid_point_u].y, sum.y);
```
remove unused code

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+               if( grid_point_u >= 0 && grid_point_u < gridsize && \
+                    grid_point_v >= 0 && grid_point_v < gridsize ) 
+	        {
+                       d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].x+= sum.x;
+                       d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].y+= sum.y;   
```
I (still) don't understand this step. As far as I can tell, at this point in the code, `sum` is always `(0, 0)`.
It's initialized as `(0, 0)` at the start of each `kk` iteration, and we shouldn't be entering this `if` conditional from one polarization to the next.
What is this supposed to be doing?

adampbeardsley commented on src/XGrid.cu:
```diff
@@ -0,0 +1,260 @@
+/* 
+
+Implements the grid-multiplication onto a GPU using CUDA. 
+
+*/
```
Missing the boiler plate stuff at the start of other .cu files. Do we need that?

Also, please give a more descriptive explanation of what this module does.

adampbeardsley commented on src/XGrid.cu:
```diff
@@ -0,0 +1,260 @@
+};
+
+template<typename In, typename Out>
+__global__ void XGrid_kernel(int npol, int gridsize, int nbatch, int nchan,
```
It doesn't appear that `gridsize`, `nbatch`, or `nchan` are used

HaraKrish replied:
Yes they do not have use in the final grid multiplication but having or removing them doesn't change or affect performance in anyway. These can be useful while performing checks (I was using them) at times to know if I'm accessing the right memory location etc..

HaraKrish replied:
Unused variable arguments have been eliminated.

adampbeardsley commented on src/XGrid.cu:
```diff
@@ -0,0 +1,260 @@
+		yy[tid_x] = d_in[bid+i%2*pol_skip+tid_x];
+	        
+		d_out[bid2+i*pol_skip+tid_x].x += xx[tid_x].x*yy[tid_x].x + xx[tid_x].y*yy[tid_x].y;  
+	       	d_out[bid2+i*pol_skip+tid_x].y += xx[tid_x].y*yy[tid_x].x - xx[tid_x].x*yy[tid_x].y;
```
Just want to double check... this function appears to be doing `x * conj(y)`. I only want to clarify because the `Complexfcma` function in `VGrid.cu` does `conj(x) * y`. Are these both correct, and what we want to be doing?

adampbeardsley replied:
`Complexfcma` is used to grid voltage data, and it's doing conj(kernel) * voltage. Here we are forming cross-pols of the entire grid (XX, XY, YX, YY), where it's convention that the second is conjugated. So this is correct.

adampbeardsley commented on src/aCorr.cu:
```diff
@@ -0,0 +1,358 @@
+#include "utils.hpp"
+#include "cuda.hpp"
+#include "cuda/stream.hpp"
+//#include <complex.h>
```
remove unused code

adampbeardsley commented on src/aCorr.cu:
```diff
@@ -0,0 +1,358 @@
+
+#define tile 256 // Number of threads per thread-block
+
+/********************/
+/* CUDA ERROR CHECK */
+/********************/
```
Why are these checks happening here but not in your other modules? Should they happen everywhere, or is this one special for some reason?

jaycedowell replied:
They seem similar to the `BF_CHECK_CUDA_EXCEPTION` and `BF_CHECK_CUDA` macros in `cuda.hpp`.

adampbeardsley commented on src/aCorr.cu:
```diff
@@ -0,0 +1,358 @@
+};
+
+template<typename In, typename Out>
+__global__ void ACorr(int nbaseline, int npol, int nbatch, int nchan,
```
`nbaseline`, `nbatch`, and `nchan` appear to be unused.

adampbeardsley commented on src/aCorr.cu:
```diff
@@ -0,0 +1,358 @@
+        In* xx = reinterpret_cast<In *>(shared);
+        In* yy= xx + blk_x; 	
+
+        // Access pattern is such that coaelescence is achieved both for read and writes to global and shared memory
```
it seems like the only difference between `ACorr` and `XGrid_kernel` is a slightly different calculation for `bid` and `bid2`. Can these be combined into a single kernel?

HaraKrish replied:
> it seems like the only difference between `ACorr` and `XGrid_kernel` is a slightly different calculation for `bid` and `bid2`. Can these be combined into a single kernel?

These have to be separate because ACorr's input is essentially the raw X and Y voltages to estimate the autocorrelation for every antenna for each frequency channel, time-stamp and polarization. So the input and output sequence(or array shapes) are different from that of XGrid which takes the X and Y uv-grids as input.

adampbeardsley reviewed (commented):
I've given my pass through, and it looks like Jayce also looked through.
@HaraKrish  there are a number of requested changes, some clarifications, and a few requests for more documentation. Please have a look, and don't forget about the `nchan` test you were going to do.

HaraKrish commented on python/bifrost/VGrid.py:
```diff
@@ -0,0 +1,52 @@
+        _check( _bf.bfVGridSetKernels(self.obj, 
+                                       asarray(kernels).as_BFarray()) )
+    def execute(self, idata, odata):
+        # TODO: Work out how to integrate CUDA stream
```
How CUDA streams are handled in Bifrost is something we need to understand.

jaycedowell replied:
It should be that each block in Bifrost gets its own CUDA stream and everything in that block uses that stream.

HaraKrish committed:
Update chips.hpp

replaced 128 with nchan. Tested and looks alright.

jaycedowell committed:
Cleanup.

jaycedowell committed:
Update src/XGrid.cu

jaycedowell committed:
Update src/aCorr.cu

jaycedowell commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+	BF_ASSERT_EXCEPTION(_y != NULL, BF_STATUS_INVALID_STATE);
+	BF_ASSERT_EXCEPTION(_z != NULL, BF_STATUS_INVALID_STATE);
+        BF_ASSERT_EXCEPTION(_kernels != NULL, BF_STATUS_INVALID_STATE);
+        BF_ASSERT_EXCEPTION(out->dtype == BF_DTYPE_CF32 \
+                                          || BF_DTYPE_CF64, BF_STATUS_UNSUPPORTED_DTYPE);
```
```suggestion
        BF_ASSERT_EXCEPTION((out->dtype == BF_DTYPE_CF32) \
                                          || (out->dtype == BF_DTYPE_CF64), BF_STATUS_UNSUPPORTED_DTYPE);
```

jaycedowell committed:
Apply suggestions from code review

jaycedowell commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+#define MIN_BLOCKS_PER_MP     4
+
+struct __attribute__((aligned(1))) nibble2 {
+    // Yikes!  This is dicey since the packing order is implementation dependent!  
```
Yes and no.  Yes, we should address it since in Bifrost changing the endianess of a `ci4` values changes the ordering of real and complex within the byte.  No, we shouldn't since I don't know that the other endianess is ever used for LWA and putting in the Bifrost endianess check would be kind of a pain.

HaraKrish committed:
Update XGrid.cu

HaraKrish committed:
Update XGrid.cu

HaraKrish committed:
Update aCorr.cu

HaraKrish committed:
Update aCorr.cu

HaraKrish committed:
Update VGrid.cu

Unused variable arguments and commented lines of code in the function VGrid_kernel were removed. Since the function definition is local these changes will not change the working of EPIC in anyway but can be tested if strictly required.

HaraKrish committed:
Update XGrid.cu

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,545 @@
+	    {
+          
+                d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].x+= sum.x;
+                d_out[grid_s + pol*gridsize*gridsize + gridsize*int(grid_point_v) + int(grid_point_u)].y+= sum.y
```
There's a syntax error on this line. We need to test the code before merging.

HaraKrish committed:
Update VGrid.cu

adampbeardsley commented on src/VGrid.cu:
```diff
@@ -0,0 +1,554 @@
+          if (!(myGridU == grid_point_u && myGridV == grid_point_v)) 
+	  //if(!((myGridU<grid_point_u+0.5)&&(myGridU>=grid_point_u-0.5)&&(myGridV<grid_point_v+0.5)&&(myGridU>=grid_point_u-0.5)))
+	   { // Atomic add to grid is now removed from this kernel after careful examining and consideration for race-conditions
+               if( grid_point_u >= 0 && grid_point_u < gridsize && \
+                    grid_point_v >= 0 && grid_point_v < gridsize ) 
```
If this doesn't hold, then you're outside the entire grid, so _yes_ we want to drop that data.

adampbeardsley committed:
clarify a conjugation

adampbeardsley commented:
Hari took some test data on Jun 9. It looks reasonable:
<img width="648" alt="image" src="https://user-images.githubusercontent.com/4752917/173095256-166dd5d6-741e-400e-8c79-e7aa26205a6f.png">
This looks reasonably like the LWA-TV grab. Note this image is about an hour off from the EPIC image because that data isn't in the rerun video.
<img width="659" alt="image" src="https://user-images.githubusercontent.com/4752917/173102971-11360eea-21b1-4098-8b43-11a957ff9563.png">


Waterfalls of some pixels look reasonable too. First is at the center of the image, second is on Cas A.
<img width="701" alt="image" src="https://user-images.githubusercontent.com/4752917/173103395-cb8341ef-0bc5-4fbe-b3a1-92a17aaf4349.png">

<img width="668" alt="image" src="https://user-images.githubusercontent.com/4752917/173103469-e2215b8e-424d-4802-b1a6-c67c8970680f.png">

mkolopanis reviewed (approved):
Recent data is looking good.
